File: prototype/collection-health.js
/* Read-only, no sample fallback and no untrusted HTML interpolation. */
(() => {
  'use strict';
  const rows = document.querySelector('#health-rows');
  const state = document.querySelector('#query-state');
  const next = document.querySelector('#next');
  let cursor = null;
  let controller = null;
  let generation = 0;
  const statuses = {selected: '선택', unreadable: '읽기 실패', enumeration_error: '탐색 오류', disabled: '비활성',
    unsupported_direct_channel: '미지원 채널', unsafe_name: '이름 제외', unsafe_path: '경로 제외', missing: '경로 없음',
    reparse_point: '링크 제외', symlink: '링크 제외', binary_archive_or_secret: '비밀·문서·바이너리 제외',
    unsupported_encoding: '미지원 인코딩', unsupported_encoding_or_binary: '인코딩·바이너리 제외',
    binary: '바이너리 제외', empty_pending: '빈 파일 대기', policy_excluded: '정책 제외'};
  const time = value => new Date(value).toLocaleString('ko-KR', {timeZone: 'Asia/Seoul'}) + ' (KST)';
  function cell(row, value) { const td = document.createElement('td'); td.textContent = String(value); row.append(td); return td; }
  async function load(after = null) {
    controller?.abort(); controller = new AbortController();
    const active = controller;
    const request = ++generation;
    const timer = setTimeout(() => active.abort(), 15000);
    rows.replaceChildren(); next.disabled = true; cursor = null; state.textContent = '조회 중';
    try {
      const response = await fetch('/api/agents/health' + (after ? '?cursor=' + encodeURIComponent(after) : ''),
        {signal: active.signal, credentials: 'same-origin', cache: 'no-store'});
      if (!response.ok) { const error = new Error('request failed'); error.login = response.status === 401; throw error; }
      const data = await response.json();
      if (request !== generation) return;
      if (!Array.isArray(data.rows) || data.rows.length > 50) throw new Error('잘못된 보고 응답입니다.');
      for (const item of data.rows) {
        const tr = document.createElement('tr');
        cell(tr, (item.host || '알 수 없음') + ' / ' + (item.organization || '-'));
        cell(tr, ({recent: '최근 수신', stale: '보고 지연', unknown: '판단 불가'})[item.report_state] || '판단 불가');
        cell(tr, `${item.selected} / ${item.excluded} / ${item.errors}`);
        cell(tr, time(item.generated_at) + ' / ' + time(item.received_at));
        cell(tr, `${item.delay_seconds ?? '?'}초 / v${item.policy_version}${item.clock_warning ? ' (시계 확인)' : ''}`);
        const td = cell(tr, '');
        const details = document.createElement('details'); const summary = document.createElement('summary');
        summary.textContent = `${item.sources.length}개 보기 (생략 ${item.omitted_sources}개)`;
        details.append(summary);
        for (const source of item.sources) {
          const line = document.createElement('p');
          const id = document.createElement('code'); id.textContent = source.id.slice(0, 12); id.title = source.id;
          line.append(id, ' : ' + (statuses[source.status] || '알 수 없음')); details.append(line);
        }
        td.append(details); rows.append(tr);
      }
      cursor = data.next_cursor; next.disabled = !cursor;
      state.textContent = data.rows.length ? `${data.rows.length}개 보고 · 큐 상태 / 전송 오류: 미측정` : '수신된 보고가 없습니다. 에이전트 미설치·구버전·전송 지연을 확인하세요.';
    } catch (error) {
      if (request !== generation) return;
      rows.replaceChildren(); cursor = null; next.disabled = true;
      state.textContent = error.name === 'AbortError' ? '조회 시간 초과. 다시 시도하세요.' : error.login ? '관리자 로그인이 필요합니다.' : '조회 실패: 서버·권한·보고 형식을 확인하세요.';
    } finally { clearTimeout(timer); }
  }
  document.querySelector('#refresh').addEventListener('click', () => load());
  next.addEventListener('click', () => { if (cursor) load(cursor); });
  load();
})();

File: src/cloud_soc/portal/collection_health.py
"""Authenticated source-discovery reports, deliberately separate from log counts."""
from datetime import datetime, timedelta, timezone
import json
import re

from cloud_soc.portal.agent_status import iso, parse_time, encode_cursor
from cloud_soc.portal.privacy import display

STATUSES = {'selected', 'unreadable', 'enumeration_error', 'disabled', 'unsupported_direct_channel',
            'unsafe_name', 'unsafe_path', 'missing', 'reparse_point', 'symlink',
            'binary_archive_or_secret', 'unsupported_encoding', 'unsupported_encoding_or_binary',
            'binary', 'empty_pending', 'policy_excluded'}


def project(source, key, now):
    report = source.get('cloud_soc', {}).get('discovery', {})
    if report.get('schema') != 1:
        raise ValueError('Unsupported health schema')
    values = {}
    for name in ('selected', 'excluded', 'errors', 'total', 'policy_version'):
        value = report.get(name)
        if type(value) is not int or not 0 <= value <= 1000000000:
            raise ValueError('Invalid health counters')
        values[name] = value
    if values['selected'] + values['excluded'] + values['errors'] != values['total']:
        raise ValueError('Inconsistent health counters')
    entries = report.get('sources')
    if not isinstance(entries, list) or len(entries) > 200 or len(entries) > values['total']:
        raise ValueError('Invalid health sources')
    sources = []
    for entry in entries:
        if (not isinstance(entry, dict) or not isinstance(entry.get('id'), str)
                or not re.fullmatch('[a-f0-9]{64}', entry['id']) or entry.get('status') not in STATUSES):
            raise ValueError('Invalid health source')
        sources.append({'id': entry['id'], 'status': entry['status']})
    generated = parse_time(report.get('generated_at'))
    received = parse_time(source.get('event', {}).get('ingested'))
    if not generated or not received:
        raise ValueError('Missing health times')
    delay = (received - generated).total_seconds()
    age = (now - received).total_seconds()
    return {**values, 'sources': sources, 'omitted_sources': values['total'] - len(sources),
            'agent_id': display(key.get('agent_id')), 'organization': display(key.get('organization')),
            'host': display(source.get('host', {}).get('name')), 'generated_at': iso(generated),
            'received_at': iso(received), 'delay_seconds': round(delay) if delay >= 0 else None,
            'clock_warning': delay < 0 or age < -60,
            'report_state': 'unknown' if age < -60 else 'recent' if age <= 300 else 'stale',
            'queue_state': 'unknown', 'transport_state': 'unknown', 'source_success': 'not_measured'}


def snapshot(client, after=None, now=None):
    now = now or datetime.now(timezone.utc)
    composite = {'size': 50, 'sources': [
        {'organization': {'terms': {'field': 'organization.id', 'missing_bucket': True}}},
        {'agent_id': {'terms': {'field': 'agent.id'}}},
        {'agent_type': {'terms': {'field': 'agent.type', 'missing_bucket': True}}},
    ]}
    if after:
        composite['after'] = after
    result = client.options(request_timeout=5, max_retries=0).search(
        index='soc-agent-health-*', size=0, timeout='4s', allow_no_indices=True,
        allow_partial_search_results=False, expand_wildcards='open',
        query={'range': {'event.ingested': {'gte': iso(now - timedelta(days=30)), 'lte': iso(now + timedelta(minutes=1))}}},
        aggs={'agents': {'composite': composite, 'aggs': {'latest': {'top_hits': {
            'size': 1, 'sort': [{'event.ingested': 'desc'}],
            '_source': ['host.name', 'cloud_soc.discovery', 'event.ingested']}}}}})
    if result.get('timed_out') or result.get('_shards', {}).get('failed', 0):
        raise ValueError('Incomplete health response')
    if result.get('_shards', {}).get('total') == 0:
        groups = {}
    else:
        groups = result['aggregations']['agents']
    rows = [project(bucket['latest']['hits']['hits'][0]['_source'], bucket['key'], now)
            for bucket in groups.get('buckets', [])]
    cursor = groups.get('after_key') if len(rows) == 50 else None
    payload = {'rows': rows, 'checked_at': iso(now), 'next_cursor': encode_cursor(cursor) if cursor else None,
               'window_days': 30, 'identity': 'agent_reported_not_attested'}
    encoded = json.dumps(payload, ensure_ascii=False).encode('utf-8')
    if len(encoded) > 2 * 1024 * 1024:
        raise ValueError('Health response too large')
    return encoded
